CDAP-17981: Make FileLocalizer to download files via AppFab instead of from HDFS

wyzhang opened the pull request:
Why:
FileLocalizer is used to localizing files needed to run twill application in k8s env.
The file is most likely in an external distributed file system (e.g. GCS in GCP)
The pod to run twill application may or may not have permission to access distributed
file system. So this change switches the localization to downloading from AppFab
(essentially make AppFab proxy the file content to twill application pod)

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileLocalizer.java:
```diff
@@ -63,17 +63,22 @@ public void run(String[] args) throws Exception {
+      }
     } else {
-      runtimeConfigLocation = new LocalLocationFactory().create(new File(uri).toURI());
+      try (InputStream  is = context.openHttpURLConnection(fileDownloadURLPath(uri)).getInputStream()) {
```
Remove extra space before `is`

chtyim replied:
You gotta disconnect the URLConnection as well

wyzhang replied:
Removed the extra space.

For disconnect, intentionally omitted it so underlying socket can be reused.
https://stackoverflow.com/a/11533423

But if you think we should always disconnect, can change to that.

chtyim replied:
The stackoverflow is a bit misleading. Since we close the InputStream here, calling disconnect will allow the connection to be reused (1st point in the SO).

Calling disconnect is for the case that if the response is non-200/300, and we don't read/close the Error stream, calling disconnect will make sure the connection won't get reused and the socket get closed correct (without leaking resources). Non connection reused is what we wanted in this case.

Alternatively, if you want to reused a socket after the response is non-200/300, you have to read and close the Error stream (`getErrorStream`).

wyzhang replied:
Changed to disconnecting upon closing stream. PTAL

chtyim commented on cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/FileLocalizer.java:
```diff
@@ -87,13 +92,14 @@ public void run(String[] args) throws Exception {
-          expand(location, targetPath);
-        } else {
-          copy(location, targetPath);
+        try (InputStream is = context.openHttpURLConnection(fileDownloadURLPath(localFile.getURI())).getInputStream()) {
```
Disconnect the URLConnection after use.

wyzhang replied:
See above response.

wyzhang replied:
Same here. Fixed.

wyzhang committed:
CDAP-17981: Make FileLocalizer to load files via AppFab instead of directly from HDFS.

Why:
FileLocalizer is used to localizing files needed to run twill application in k8s env.
The file is most likely in an external distributed file system (e.g. GCS in GCP)
The pod to run twill application may or may not have permission to access distributed
file system. So this change swtiches the localization to downloading from AppFab
(essentially make AppFab proxy the file content to twill application pod)
